### PR TITLE
Porting of OrcidPublicationDataProvider

### PR DESCRIPTION
**Pull request by LucaGiamminonni**
## References
Includes https://github.com/DSpace/DSpace/pull/8266

**Unique code in this PR (~1,600 lines of code) can be found via this diff:**
https://github.com/DSpace/DSpace/compare/4Science:CST-5587...4Science:CST-5799

## Description
This PR introduces the external data provider OrcidPublicationDataProvider to import publications from ORCID by ORCID id

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+import org.orcid.jaxb.model.v3.release.record.summary.WorkSummary;
+import org.orcid.jaxb.model.v3.release.record.summary.Works;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Implementation of {@link ExternalDataProvider} that search for all the works
+ * of the profile with the given orcid id that hava a source other than
+ * DSpaceCris. The id of the external data objects returned by the methods of
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java:71]: Comment mentions DSpaceCris, when I think it just means "DSpace"

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+ * this class is the concatenation of the orcid id and the put code associated
+ * with the publication, separated by ::
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java:73]: Can we add an example output in the JavaDocs here?  Are you saying the ID format with be `[orcid]::[putcode]`?  What's an example Put Code?

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        return findWorks(orcid, start, limit).stream()
+            .map(work -> convertToExternalDataObject(orcid, work))
+            .collect(Collectors.toList());
+    }
+
+    private boolean isInvalidIdentifier(String id) {
+        return StringUtils.isBlank(id) || id.split("::").length != 2;
+    }
+
+    private List<Work> findWorks(String orcid, int start, int limit) {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java:133]: Starting here, there are a large number of Private methods which have no comments or JavaDocs.  Please describe them, as it's not always clear what they are doing based on params.  E.g. is this returning "paginated" Works?

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    private WorkBulk getWorkBulk(String orcid, List<String> putCodes) {
+        if (orcidConfiguration.isApiConfigured()) {
+            String accessToken = getAccessToken(orcid);
+            return orcidClient.getWorkBulk(accessToken, orcid, putCodes);
+        } else {
+            return orcidClient.getWorkBulk(orcid, putCodes);
+        }
+    }
+
+    private String getAccessToken(String orcid) {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidPublicationDataProvider.java:193]: Obviously the code around access tokens will need to be updated in this PR

File: dspace-api/src/test/data/dspaceFolder/assetstore/orcid-works/work-277871.xml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (tdonohue) [inline dspace-api/src/test/data/dspaceFolder/assetstore/orcid-works/work-277871.xml:1]: As noted in other PRs, these test data files should be moved to `src/test/resources` alongside the IT which uses them. That way they are easily findable next to the IT.

File: dspace-api/src/test/java/org/dspace/external/provider/impl/OrcidPublicationDataProviderIT.java [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+/**
+ * Integration tests for {@link OrcidPublicationDataProvider}.
+ *
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ *
+ */
+public class OrcidPublicationDataProviderIT extends AbstractIntegrationTestWithDatabase {
+
+    private static final String BASE_XML_DIR_PATH = "./target/testing/dspace/assetstore/orcid-works/";
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/external/provider/impl/OrcidPublicationDataProviderIT.java:58]: Once the test files are moved, this hardcoded path isn't needed anymore. You can just get the resources by name.

File: dspace-api/src/test/java/org/dspace/external/provider/impl/OrcidPublicationDataProviderIT.java [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+
+        orcidClientMock = mock(OrcidClient.class);
+        orcidClient = dataProvider.getOrcidClient();
+
+        dataProvider.setReadPublicAccessToken(null);
+        dataProvider.setOrcidClient(orcidClientMock);
+
+        originalClientId = orcidConfiguration.getClientId();
+        orcidConfiguration.setClientId("DSPACE-CRIS-CLIENT-ID");
+        orcidConfiguration.setClientSecret("DSPACE-CRIS-CLIENT-SECRET");
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/external/provider/impl/OrcidPublicationDataProviderIT.java:105]: Please change the DSPACE-CRIS example settings here to just say DSPACE (or something else)

**Review by tdonohue — COMMENTED** [review]
@LucaGiamminonni : Obviously this PR isn't in a testable state right now (which is OK, I understand why it isn't).  But, I wanted to give this an early code review just so that it has some feedback.  Overall the code looks reasonable at a glance, but there is some minor cleanup (JavaDocs missing, test files in the wrong place, etc) which would be useful in this PR.

**Review by MarieVerdonck — CHANGES_REQUESTED** [review]
Code looks good and seems to be working well, thanks!
One note though, when using an ORCID id that is not valid or not found in the (public) orcid api, the externalsources endpoint gives a 500 error instead of indicating what the issue may be.

For example `/api/integration/externalsources/orcidWorks/entries?query=test` => 500 error, log: `<developer-message>404 Not Found: The resource was not found. Full validation error: ORCID iD test not found</developer-message>` => Better would be a 400 illegal argument / bad request with message that it needs to a valid orcid id.
The same happens when you do use a valid orcid (format), but is not found in the public orcid (`https://pub.orcid.org/v3.0/{orcidID}/works`); eg `query=0000-0001-9227-636X`

File: dspace/config/crosswalks/orcid/mapConverter-orcid-to-dspace-language-code.properties [hunk 8]
```diff
@@ -0,0 +1,8 @@
+en = en
```
Comment (tdonohue) [inline dspace/config/crosswalks/orcid/mapConverter-orcid-to-dspace-language-code.properties:1]: Is this mapping necessary? It looks like the same value is always in the key and in the value?  If it is necessary, I think it needs a comment to describe it's usage.

File: dspace/config/crosswalks/orcid/mapConverter-orcid-to-dspace-publication-type.properties [hunk 9]
```diff
@@ -0,0 +1,11 @@
+website = Other
```
Comment (tdonohue) [inline dspace/config/crosswalks/orcid/mapConverter-orcid-to-dspace-publication-type.properties:1]: Please add a comment at the top of this file to describe the key/value.  I'm assuming this is essentially the opposite of `mapConverter-dspace-to-orcid-publication-type.properties`?

**Review by tdonohue — COMMENTED** [review]
@LucaGiamminonni :  I gave this new Import Source a quick test today, and it's working as expected.  I can enter in a valid ORCID (for sandbox.orcid.org) and see that person's works returned...at which point I can import one of those works into DSpace.

I also looked at the code (which is mostly new ITs, and only really one new class along with minor updates).  It looks good overall, but I have a few minor inline comments below.

Overall, this looks good & I'm basically a +1.  However, I'd want to re-review/retest after #8266 is merged (hopefully tomorrow).  So, it's possible this could squeeze into 7.3 as it's quite small once the other PR is merged.

**Review by pnbecker — APPROVED** [review]
I gave this a review using the link provided comparing two branches in 4Science's repository. The code looks good to me. I have not tested it, but @tdonohue hast tested it yesterday.

:+1:

**Review by MarieVerdonck — CHANGES_REQUESTED** [review]
Thanks @LucaGiamminonni 
A non valid ORCID format now indeed gives informative log + 404 error, but not yet the case for valid format but non found ORCID:

`/server/api/integration/externalsources/orcidWorks/entries?query={nonExistentORCIDcorrectFormat}` (eg `query=0000-0003-4907-5057`) still gives a 500 error

This is coming from the ORCID API error:

```
org.dspace.orcid.exception.OrcidClientException: <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<error xmlns="http://www.orcid.org/ns/error">
    <response-code>404</response-code>
    <developer-message>404 Not Found: The resource was not found. Full validation error: ORCID iD 0000-0003-4907-5057 not found</developer-message>
    <user-message>The resource was not found.</user-message>
    <error-code>9016</error-code>
    <more-info>https://members.orcid.org/api/resources/troubleshooting</more-info>
</error>
```

=> I would pass along this error message from the ORCID API to the DSpace log and return the same status (404) when using a ORCID ID that is valid format but doesn't resolve.

**Comment by tdonohue** [comment]
@LucaGiamminonni : When you get a chance can you rebase this on latest `main`?  Also give a look to @MarieVerdonck 's feedback to see if it's something that can be addressed quickly?  I agree it'd be ideal to return a 404 if the ORCID API also returns a 404...but if that's too complex we could also log a "known bug" to address later.

**Comment by abollini** [comment]
a `404`  from ORCID should indeed lead to an empty response as the query has produced no results. It is not a 404 from the DSpace perspective as our "external source search" endpoint exists and works properly, it will return a 0 size list with a `200` return code. @LucaGiamminonni  is working on this change right now.

We should also revisit the `400` return code that is currently returned in the case an external provider throw an `IllegalArgumentException`, it would be more appropriate to convert it to a `422` as from the perspective of a rest client the dspace endpoint has been called "in a syntactically" correctly way but the provided query doesn't match additional business criteria. Unfortunately, the external source endpoint doesn't document the return code in the contract so we will suggest to make this last change (400 -> 422) in a separate PR for 7.4 associated with a REST Contract.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @LucaGiamminonni !  Retested today and it works for me. The code also looks good and my feedback has been addressed.

Merging this immediately as it's at +2.

NOTE: Please disregard the test failures in this PR.  They existed on `main` and have been fixed by #8368 .  I've run the new test added by this PR locally & it succeeds.

**Pull request merged**